### PR TITLE
Update dummy fields within the ref by auto-generating them from the source assembly.

### PR DESCRIPTION
**Pull request by ahsonkhan**
From https://github.com/dotnet/corefx/issues/39652#issuecomment-513915482

Picked out the "dummy field" related changes when auto-generated the reference assembly from source (from ~https://github.com/dotnet/corefx/commit/fa4a2d7a453f3c947e2277e1058457601452feb9~ https://github.com/ahsonkhan/corefx/commit/ceec15160b977fa4850c3974f3b015bcfdbc1b34 which contains a lot of other changes that we should also consider at some point, like param renames or missing/mismatching API surface).

~Some of them don't make sense to me.~

cc @stephentoub

### CODE REVIEW

File: src/System.Diagnostics.StackTrace/ref/System.Diagnostics.StackTrace.cs [hunk 1]
```diff
@@ -172,7 +172,7 @@ public enum SymAddressKind
     }
     public readonly partial struct SymbolToken
     {
-        private readonly int _dummyPrimitive;
+        private readonly object _dummy;
```
Comment (ahsonkhan) [inline src/System.Diagnostics.StackTrace/ref/System.Diagnostics.StackTrace.cs]: I don't know why some of these changed from primitive fields to reference type fields.

For instance, this is literally a struct wrapping an int:
https://github.com/dotnet/corefx/blob/16f3eef83d6df4c9dbc0e2992dafbc9553db363d/src/System.Diagnostics.StackTrace/src/System/Diagnostics/SymbolStore/SymbolToken.cs#L9

cc @ericstj

File: src/System.Reflection.Metadata/src/System/Reflection/Metadata/MetadataReaderProvider.cs [hunk 2]
```diff
@@ -243,7 +243,7 @@ public void Dispose()
         /// <exception cref="PlatformNotSupportedException">The current platform is big-endian.</exception>
         /// <exception cref="IOException">IO error while reading from the underlying stream.</exception>
         /// <exception cref="ObjectDisposedException">Provider has been disposed.</exception>
-        public unsafe MetadataReader GetMetadataReader(MetadataReaderOptions options = MetadataReaderOptions.Default, MetadataStringDecoder utf8Decoder = null)
+        public unsafe MetadataReader GetMetadataReader(MetadataReaderOptions options = MetadataReaderOptions.ApplyWindowsRuntimeProjections, MetadataStringDecoder utf8Decoder = null)
```
Comment (ahsonkhan) [inline src/System.Reflection.Metadata/src/System/Reflection/Metadata/MetadataReaderProvider.cs]: There was a source/ref assembly mismatch, so I fixed the source based on the documentation:
https://github.com/dotnet/corefx/blob/16f3eef83d6df4c9dbc0e2992dafbc9553db363d/src/System.Reflection.Metadata/src/System/Reflection/Metadata/MetadataReaderOptions.cs#L15-L20

File: src/System.Security.Cryptography.Cng/ref/System.Security.Cryptography.Cng.cs [hunk 3]
```diff
@@ -199,6 +199,7 @@ public enum CngKeyUsages
     public partial struct CngProperty : System.IEquatable<System.Security.Cryptography.CngProperty>
     {
         private object _dummy;
+        private int _dummyPrimitive;
```
Comment (ahsonkhan) [inline src/System.Security.Cryptography.Cng/ref/System.Security.Cryptography.Cng.cs]: cc @bartonjs

File: src/System.Threading/ref/System.Threading.cs [hunk 4]
```diff
@@ -186,7 +186,7 @@ public static partial class LazyInitializer
     }
     public partial struct LockCookie
     {
-        private int _dummyPrimitive;
+        private object _dummy;
```
Comment (ahsonkhan) [inline src/System.Threading/ref/System.Threading.cs]: Similarly here, not sure why this change was made:
https://github.com/dotnet/corefx/blob/16f3eef83d6df4c9dbc0e2992dafbc9553db363d/src/System.Threading/src/System/Threading/LockCookie.cs#L13-L17

File: src/System.Reflection.Metadata/ref/System.Reflection.Metadata.cs [hunk 5]
```diff
@@ -85,6 +85,7 @@ namespace System.Reflection.Metadata
     public readonly partial struct ArrayShape
     {
         private readonly object _dummy;
+        private readonly int _dummyPrimitive;
```
Comment (ahsonkhan) [inline src/System.Reflection.Metadata/ref/System.Reflection.Metadata.cs]: cc @nguerrera, @tmat to review the Reflection.Metadata changes. 

Spot checked and seems to make sense:
https://github.com/dotnet/corefx/blob/16f3eef83d6df4c9dbc0e2992dafbc9553db363d/src/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeSystem/AssemblyFile.cs#L11-L14
https://github.com/dotnet/corefx/blob/16f3eef83d6df4c9dbc0e2992dafbc9553db363d/src/System.Reflection.Metadata/src/System/Reflection/Metadata/Signatures/CustomAttributeNamedArgument.cs#L7-L12
https://github.com/dotnet/corefx/blob/16f3eef83d6df4c9dbc0e2992dafbc9553db363d/src/System.Reflection.Metadata/src/System/Reflection/Metadata/TypeSystem/ExportedType.cs#L9-L14

**Comment by ahsonkhan** [comment]
I left out `System.Runtime.WindowsRuntime.cs` for now until we have clarity on the expectation. I have the same question for the types it exposes like Windows.UI.Color which I expect to have primitive fields (but auto-gen changes to object field):
https://github.com/dotnet/corefx/blob/16f3eef83d6df4c9dbc0e2992dafbc9553db363d/src/System.Runtime.WindowsRuntime/src/System/Windows/UI/Color.cs#L35-L40

File: src/System.Reflection.Metadata/src/System/Reflection/Metadata/MetadataReaderProvider.cs (hunk 2 above)
Comment (nguerrera) [reply src/System.Reflection.Metadata/src/System/Reflection/Metadata/MetadataReaderProvider.cs]: These have the same value, but the intent was to use Default. I think this came up last time.

Comment (ahsonkhan) [reply src/System.Reflection.Metadata/src/System/Reflection/Metadata/MetadataReaderProvider.cs]: Would this still be considered a breaking change?

Also, looking at the documentation (https://docs.microsoft.com/en-us/dotnet/api/system.reflection.metadata.metadatareaderoptions?view=netcore-2.2), it seems contradictory to use `Default` when the overload takes `MetadataReaderOptions`, which is the case here.
```text
The options that are used when a MetadataReader is obtained via an overload that does not take a MetadataReaderOptions argument.
```

Comment (nguerrera) [reply src/System.Reflection.Metadata/src/System/Reflection/Metadata/MetadataReaderProvider.cs]: I think that comment dates from before any optional params were used or allowed. Default and ApplyWindowsRuntimeProjection have the same value. They are indistinguishable at runtime. It is the default. So this is a styke change onky. 

The comment could use updating. But the idea remains that you get default option when yoy don't pass your own value. 

Mind you, I think maybe the intent was that Default could change to include other options, but that could be breaking, so maybe it wasn't the best idea.

Comment (nguerrera) [reply src/System.Reflection.Metadata/src/System/Reflection/Metadata/MetadataReaderProvider.cs]: It would have been intended to use Default inthe ref sources too. There was no ref assembly when this library was first written. The problem is that the tooling operates on binaries not source so it can't tell which spelling the source used.

File: src/System.Diagnostics.StackTrace/ref/System.Diagnostics.StackTrace.cs (hunk 1 above)
Comment (stephentoub) [reply src/System.Diagnostics.StackTrace/ref/System.Diagnostics.StackTrace.cs]: Yeah, something looks wrong here...

Comment (ericstj) [reply src/System.Diagnostics.StackTrace/ref/System.Diagnostics.StackTrace.cs]: /cc @safern Perhaps this is due to being unable to resolve System.Private.Corelib so GenAPI can't determine that System.Int32 is actually a value type.

File: src/System.Reflection.Metadata/ref/System.Reflection.Metadata.cs [hunk 6]
```diff
@@ -1684,7 +1733,7 @@ public sealed partial class MetadataReaderProvider : System.IDisposable
         public unsafe static System.Reflection.Metadata.MetadataReaderProvider FromPortablePdbImage(byte* start, int size) { throw null; }
         public static System.Reflection.Metadata.MetadataReaderProvider FromPortablePdbImage(System.Collections.Immutable.ImmutableArray<byte> image) { throw null; }
         public static System.Reflection.Metadata.MetadataReaderProvider FromPortablePdbStream(System.IO.Stream stream, System.Reflection.Metadata.MetadataStreamOptions options = System.Reflection.Metadata.MetadataStreamOptions.Default, int size = 0) { throw null; }
-        public System.Reflection.Metadata.MetadataReader GetMetadataReader(System.Reflection.Metadata.MetadataReaderOptions options = System.Reflection.Metadata.MetadataReaderOptions.ApplyWindowsRuntimeProjections, System.Reflection.Metadata.MetadataStringDecoder utf8Decoder = null) { throw null; }
+        public System.Reflection.Metadata.MetadataReader GetMetadataReader(System.Reflection.Metadata.MetadataReaderOptions options = System.Reflection.Metadata.MetadataReaderOptions.Default, System.Reflection.Metadata.MetadataStringDecoder utf8Decoder = null) { throw null; }
```
Comment (ahsonkhan) [inline src/System.Reflection.Metadata/ref/System.Reflection.Metadata.cs]: As per discussion, updated the ref instead (the underlying enum values are the same since `Default == ApplyWindowsRuntimeProjections`).

**Comment by karelz** [comment]
This PR went into master - fixing milestone to 5.0.
It was later ported into release/3.0 branch in PR #39747.
